src/dune_targets/dune_targets.ml

```diff
@@ -243,9 +243,7 @@ module Produced = struct
                 acc_filenames, Path.Local.Map.union_exn acc_dirs dir
               | _ -> Error (Unsupported_file (Path.Build.relative dir filename, kind)))
         in
-        if not (Filename.Map.is_empty filenames)
-        then Path.Local.Map.add_exn dirs local filenames
-        else dirs
+        Path.Local.Map.add_exn dirs local filenames
```

Comment:
I'm not sure for this fix, I think this code is fine. The issue is coming from somewhere else.

Reply:
Thanks a lot for the help. My impression is that indeed it matters that we add the binding `{ dir -> [] }` to the map, as even if the directory doesn't have files, it can have subdirs.

But I didn't track down exactly the part that makes the current code break.

Reply:
So indeed you are correct that the crash happens at `remove_state_files_and_subdirectories`, but my impression is that we break the invariant about source representation that the above function requires.

Reply:
I get it, that's fair point. 

The question I have now is, are we going to allow promoting empty directories ?

Reply:
Maybe the confusion is about what we call "empty directory", empty directories are OK not to promote in my view, however here the directories in question are not empty, they just have no files, however they do have other subdirectories, so definitively they need to be considered part of the sources.

Reply:
I tested the fix, it is promoting even empty directories.
This is not fixing exactly the test-case, it's fixing it indirectly by allowing an empty directory.

Reply:
Indeed after this, empty directories are promoted. I don't see why they shouldn't be promoted?

We could optimize this away, but I am not sure it is worth it.

Note that the fix doesn't exactly allow "empty directories", it allows too, crutially, directories that have no files but are not empty, such as the test case.

Reply:
Looking to this further. I agree with you, this where we need to fix the issue.

Reply:
This is one of the solution https://github.com/moyodiallo/dune/commit/b00b92e1fbc1007736677436d3b9b56be3f089fc. 

The thing is adding all the parent directories of the collected files, it will ends up with no empty directory.